src/PhenX.EntityFrameworkCore.BulkInsert/Dialect/SqlDialectBuilder.cs

```diff
@@ -122,18 +140,28 @@ protected virtual void AppendOnConflictStatement(StringBuilder sql)
         sql.AppendLine("ON CONFLICT");
     }
 
-    protected virtual void AppendConflictCondition<T>(StringBuilder sql, OnConflictOptions<T> onConflictTyped)
+    protected virtual void AppendConflictCondition<T>(StringBuilder sql, TableMetadata target, DbContext context,
+        OnConflictOptions<T> onConflictTyped)
     {
-        sql.AppendLine($"WHERE {onConflictTyped.Condition}");
+        var condition = onConflictTyped.RawWhere;
+
+        if (onConflictTyped.Where != null)
+        {
+            condition = ToSqlExpression<T>(context, target, onConflictTyped.Where);
+        }
+
+        sql.Append(condition).AppendLine();
     }
 
     /// <summary>
-    /// Get the name of the excluded column for the ON CONFLICT clause.
+    /// Get the name of the INSERTED column (data already in the table) for the ON CONFLICT clause.
     /// </summary>
-    protected virtual string GetExcludedColumnName(string columnName)
-    {
-        return $"EXCLUDED.{Quote(columnName)}";
-    }
+    protected virtual string GetInsertedColumnName(string columnName) => $"INSERTED.{Quote(columnName)}";
```

Comment:
I would either introduce constants for these two tables somewhere or inject the table name somehow, so that both terms are managed in one place